files/2-finalized/awaited-base/AwaitedIterator.ts

```diff
@@ -0,0 +1,47 @@
+import AwaitedPath from './AwaitedPath';
+import AwaitedHandler from './AwaitedHandler';
+
+export default class AwaitedIterator<TClass, T> {
+  public static creators = require('../awaited-dom/create');
+  private readonly createItemName: string;
+  private readonly getState: (instance: TClass) => IAwaitedIteratorProperties;
+  private awaitedHandler: AwaitedHandler<TClass>;
+
+  constructor(
+    createItemName: string,
+    getState: (instance: TClass) => IAwaitedIteratorProperties,
+    awaitedHandler: AwaitedHandler<TClass>,
+  ) {
+    this.createItemName = createItemName;
+    this.getState = getState;
+    this.awaitedHandler = awaitedHandler;
+  }
+
+  public iterateAttachedNodeIds(instance: TClass): IterableIterator<T> {
+    const { awaitedPath } = this.getState(instance);
+    return this.getNodeList(instance, awaitedPath.iterableIds ?? [])[Symbol.iterator]();
+  }
+
+  public async toArray(instance: TClass): Promise<T[]> {
+    const [, iterableIds] = await this.awaitedHandler.getNodeIds(instance);
+    return this.getNodeList(instance, iterableIds);
+  }
+
+  private getNodeList(instance: TClass, nodeIds: number[]): T[] {
+    if (!nodeIds) return [];
+
+    const state = this.getState(instance);
+    const awaitedPath = state.awaitedPath as AwaitedPath;
+    const awaitedOptions = state.awaitedOptions;
+    const createChild = AwaitedIterator.creators[this.createItemName];
+    return nodeIds.map(nodeId => {
+      const childPath = awaitedPath.withNodeId(nodeId);
+      return createChild(childPath, awaitedOptions) as T;
+    });
```

Comment:
Why create ALL the child classes that "might" be iterated through. Wouldn't it be much cleaner to create an actual iterator that creates the children as the list is iterated through. And if the for loop is exited then there is no need to finish creating the remainder of children.

Reply:
It seems to me that this pull request is not creating an "Awaited Iterator". It's really just implementing the concept of attachedId(s) that are generated when the object is "awaited".

NodeList's iterator implementation is distinct from "awaited", well, kinda... obviously it would use the attachedIds from state to iterate on... but if someone tries to iterate on a non-awaited NodeList class (i.e., non-attached) then the iterator could still work, it would just iterate on an empty attachedIds list... or perhaps we should throw an error inside the iterator if the object hasn't been awaited/attached.

Reply:
I started with a true iterator, and it just wouldn't trigger, so I tried to just create an array of elements instead. This approach worked, and I reasoned that 1) converting the array of numbers to wrapper objects was not very expensive, and 2) we would be dealing 99% of the time with pretty small lists, and 3) we could always improve efficiency if we needed to sometime in the future.

Regarding awaited iterator - are you commenting on confusion/unhappiness with the internally used name? If so, it's definitely evolved a ton since we talked about attaching ids, for sure. It is mostly just abstracting the primary iteration and creation of sub-items out of the base classes.

Your last point about simply iterating on an empty list was what I proposed on the phone and I (maybe mistakenly) thought you said "yes, exactly". An error could work too though. Since you can just kind of keep awaiting it and getting something different, I'm not sure which way is clearer